引入mapstruct，用于entity和dto的映射，完成获取题目接口

KamenRiderKuuga opened the pull request.

KamenRiderKuuga committed:
引入mapstruct，用于entity和dto的映射

KamenRiderKuuga committed:
完成获取题目接口

timeaftertime commented on qmoj/src/main/java/team/akina/qmoj/constants/Enums.java:
```diff
@@ -0,0 +1,36 @@
+package team.akina.qmoj.constants;
+
+public class Enums {
```
一般是每个枚举一个类，可以都放在一个包里，而不是所有类放在一个 Enums 类里

KamenRiderKuuga replied:
因为想到这样可能会产生很多文件，在C#里经常喜欢这样写，我改改

timeaftertime commented on qmoj/src/main/java/team/akina/qmoj/dto/QmojQuestionDto.java:
```diff
@@ -0,0 +1,100 @@
+/**
+ * 题目内容的dto类，沿用chenbo以来的命名风格，添加Qmoj前缀
+ */
+public class QmojQuestionDto {
```
DTO

KamenRiderKuuga replied:
这边按照[mapstruct的官方示例](https://mapstruct.org/)写的，写的时候也犹豫了一下，这边统一改成大写的吧，在语义上正确些

timeaftertime commented on qmoj/src/main/java/team/akina/qmoj/dto/QmojQuestionDto.java:
```diff
@@ -0,0 +1,100 @@
+    /**
+     * 题目的标签，中文形式
+     */
+    private List<String> topic_tags;
```
应该用驼峰命名

KamenRiderKuuga replied:
这里中邪了，用成python的命名方式了，下次部署的时候让前端同步

timeaftertime commented on qmoj/src/main/java/team/akina/qmoj/dto/QmojQuestionDto.java:
```diff
@@ -0,0 +1,100 @@
+    private List<String> topic_tags;
+
+
+    public long getId() {
```
可以直接用 Lombok 的 @Data 注解

KamenRiderKuuga replied:
我这里的插件不知道出了啥问题，没配好，感觉Lombok入侵性有点强，这里先不用Lombok了

timeaftertime replied:
波波仔已经用了Lombok，入侵行已经不是你用不用的问题了吧

timeaftertime commented on qmoj/src/main/java/team/akina/qmoj/dto/QmojQuestionDto.java:
```diff
@@ -0,0 +1,100 @@
+        this.languages = languages;
+    }
+
+    public List<String> getTopic_tags() {
```
同驼峰

KamenRiderKuuga replied:
同上

timeaftertime commented on qmoj/src/main/java/team/akina/qmoj/dto/mapper/QmojQuestionDtoMapper.java:
```diff
@@ -0,0 +1,59 @@
+     * @param question 题目详情实体
+     * @return
+     */
+    @Mappings({@Mapping(target = "platform", expression = "java(platFormConvert(question.getPlatform()))"),
```
尽量不要写 expression ，这样不方便维护
这种简单的 platform 到 getPlatform 应该是可以直接映射的

KamenRiderKuuga replied:
这里是一个枚举到一个平台类型字符串的转换，不写expression用不出来，考虑到MapStruct是会生成文件的，如果写错了是编译不通过的，这里还是继续使用了expression

timeaftertime replied:
OK，没注意到这里需要额外转换

timeaftertime commented on qmoj/src/main/java/team/akina/qmoj/dto/mapper/QmojQuestionDtoMapper.java:
```diff
@@ -0,0 +1,59 @@
+     * @param code 平台类型编码
+     * @return
+     */
+    default String platFormConvert(byte code) {
```
这应该是 Platform 这个枚举的逻辑，根据职责划分应该放在 Platform 枚举里

KamenRiderKuuga replied:
删除了此函数

timeaftertime commented on qmoj/src/main/java/team/akina/qmoj/dto/mapper/QmojQuestionDtoMapper.java:
```diff
@@ -0,0 +1,59 @@
+     * @param languages
+     * @return
+     */
+    default List<Integer> languagesConvert(String languages) {
```
应该用 static 而不是 default

KamenRiderKuuga replied:
作为在interface里的定义，这里用default比较合适，也方便写expression

timeaftertime replied:
这个是单纯的转换代码，和实例无关，应该用 static 。实例方法才定义成 default ，用于给子类覆盖

timeaftertime commented on qmoj/src/main/java/team/akina/qmoj/dto/mapper/QmojQuestionDtoMapper.java:
```diff
@@ -0,0 +1,59 @@
+     * @param tags
+     * @return
+     */
+    default List<String> tagsConvert(String tags) {
```
同上

KamenRiderKuuga replied:
同上

timeaftertime commented on qmoj/src/main/java/team/akina/qmoj/exception/DataNotFindException.java:
```diff
@@ -0,0 +1,14 @@
+/**
+ * 没有找到数据时引发的异常
+ */
+public class DataNotFindException extends CustomException {
```
CustomException  是谁取的名字……吊起来打……

KamenRiderKuuga replied:
这自定义异常，没毛病！

timeaftertime replied:
应该用体现业务含义的名字，而不是 “自定义异常”、“波波仔异常”、“GK异常”

timeaftertime commented on qmoj/src/main/java/team/akina/qmoj/exception/ExceptionHelper.java:
```diff
@@ -34,6 +34,15 @@ public Response handleLeetCodeCrawlerException(LeetCodeCrawlerException ex) {
+     */
+    @ExceptionHandler(value = {DataNotFindException.class})
+    public Response dataNotFindException(DataNotFindException ex) {
+        logger.error(Arrays.toString(ex.getStackTrace()));
```
确定这种方式能打印完整、易读的堆栈信息吗

KamenRiderKuuga replied:
完整，不易读，后续调整

timeaftertime commented on qmoj/src/main/java/team/akina/qmoj/exception/ExceptionHelper.java:
```diff
@@ -34,6 +34,15 @@ public Response handleLeetCodeCrawlerException(LeetCodeCrawlerException ex) {
+    /**
+     * 兜底的异常
+     */
+    @ExceptionHandler(value = {DataNotFindException.class})
```
全局异常处理器放在这个位置是不是不太合适，而且确定这样能生效吗，能被 Spring 管理吗

KamenRiderKuuga replied:
这样可以生效，这里注释有问题，是没有找到数据时引发的异常

timeaftertime commented on qmoj/src/test/java/team/akina/qmoj/dto/DtoConvertTest.java:
```diff
@@ -0,0 +1,52 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+@SpringBootTest
+public class DtoConvertTest {
```
同 DTO

timeaftertime commented on qmoj/src/test/java/team/akina/qmoj/dto/DtoConvertTest.java:
```diff
@@ -0,0 +1,52 @@
+    /**
+     * 获取当前时间
+     */
+    public Date getCurrentTime() {
```
获取时间的方式似乎有点复杂了，一般是 Date 加 SimpleDateFormatter

KamenRiderKuuga replied:
这里是想获取当前时间来着，原来只用new一个就行了，写习惯DateTime.Now，以为Java没有，还百度了一圈

KamenRiderKuuga committed:
解决https://github.com/qms-dev/qmoj-core/pull/5提到的问题

KamenRiderKuuga committed:
解决https://github.com/qms-dev/qmoj-core/pull/5提到的问题

KamenRiderKuuga committed:
获取所有题目摘要列表

KamenRiderKuuga committed:
统一后端题目标签字段名称为topicTags

KamenRiderKuuga committed:
统一后端题目标签字段名称为topicTags

KamenRiderKuuga committed:
今晚要重装系统，提交昨晚忘记的字段调整

KamenRiderKuuga committed:
整合Redis，使用Lettuce客户端

KamenRiderKuuga committed:
Merge branch 'dev' into dev-hanabi

KamenRiderKuuga committed:
实现提交题解以及查询判题结果功能

21kl commented on qmoj/src/main/java/team/akina/qmoj/pojo/QmojAnswer.java:
```diff
@@ -0,0 +1,46 @@
+/**
+ * 用来映射前端提交的题解
+ */
+public class QmojAnswer {
```
映射前端传参可以给类加个Param或者VO更加清晰写

KamenRiderKuuga replied:
> 映射前端传参可以给类加个Param或者VO更加清晰写

好，现在将前端参数传入归类为Param了！

21kl commented on qmoj/src/main/java/team/akina/qmoj/pojo/QmojAnswer.java:
```diff
@@ -0,0 +1,46 @@
+    /**
+     * 提交题解的语言类型
+     */
+    private int language;
```
我记得还有一个很重要的参数，叫做slug的，为啥这里前端不直接传过来，里扣提交用不到嘛？还是准备用id去找slug？

KamenRiderKuuga replied:
> 我记得还有一个很重要的参数，叫做slug的，为啥这里前端不直接传过来，里扣提交用不到嘛？还是准备用id去找slug？

现在是采用了用id找slug的方式，因为考虑到可能不止是从LeetCode来的题，这里就没对LeetCode的相关参数做太特别的处理

21kl commented on qmoj/src/test/java/team/akina/qmoj/utils/RedisTest.java:
```diff
@@ -0,0 +1,44 @@
+        answer.setJudge_type("large");
+        answer.setLang("cpp");
+        answer.setQuestion_id(1);
+        answer.setTest_mode(false);
```
这个answer类命名似乎不太规范。。最好统一都用驼峰的形式。

KamenRiderKuuga replied:
> 这个answer类命名似乎不太规范。。最好统一都用驼峰的形式。

这里是根据LeetCode需要的Json生成的pojo，这类临时的pojo现在都放在pojo这个package，因为之前出现过getter，setter命名不规范影响到第三方包进行反序列化，我感觉这里保持这样吧，如果有异议欢迎继续讨论哈

21kl commented on qmoj/src/main/java/team/akina/qmoj/controller/AnswerController.java:
```diff
@@ -0,0 +1,32 @@
+import team.akina.qmoj.service.QmojQuestionService;
+
+@RestController
+@RequestMapping("/api")
```
api前缀按道理一般是给外部系统调用的，我理解这边只是给前端调用其实没必要用api前缀。用资源名比如answer就好

KamenRiderKuuga replied:
> api前缀按道理一般是给外部系统调用的，我理解这边只是给前端调用其实没必要用api前缀。用资源名比如answer就好

搜了一圈，没有找到明确说关于这个的，我们这里外部要调用的话可以再加个`open`啥的，不过想了想，这里还是去掉`@RequestMapping("/api")`比较好，至少可以省代码。在下一次发布的时候会更新文档，并且告诉前端

KamenRiderKuuga committed:
调整前端传入参数的分类，调整api路由规范

Comment from @21kl
https://github.com/qms-dev/qmoj-core/pull/5#discussion_r549094160
https://github.com/qms-dev/qmoj-core/pull/5#discussion_r549095110

Co-Authored-By: 21kl <45235187+21kl@users.noreply.github.com>

gk008 reviewed (commented):
USER_AGENT为啥要改掉

KamenRiderKuuga reviewed (commented):
viewed

KamenRiderKuuga reviewed (commented):
viewed

KamenRiderKuuga reviewed (commented):
viewed

KamenRiderKuuga commented:
@gk008 USER_AGENT改掉是那天晚上调试查询判题结果接口的时候，一直返回一个叫`PENDING`的状态字符串，获取不到结果。我以为是LeetCode识别出来爬虫了，就不停想去尽量还原浏览器行为，就完善了一下这个USER_AGENT。事实证明对我要解决的问题没什么帮助，不过想着USER_AGENT完善点肯定没啥坏处，就保留了
